csrc/multidevice/communication.h

```diff
@@ -15,43 +15,39 @@
 
 namespace nvfuser {
 
-/*
-  This struct gathers all the parameters necessary for the
-  construction a communication
-*/
+// This struct gathers all the parameters necessary for the
+// construction a communication
 struct CommParams {
   DeviceIdxType root = -1;
-  std::vector<at::Tensor> src_bufs;
-  std::vector<at::Tensor> dst_bufs;
-  Team team; // should not have duplicate
+  bool is_root_in_mesh = true;
+  Team team; // should not have duplicates and should contain both the root and
+             // the mesh
   c10d::ReduceOp::RedOpType redOp = c10d::ReduceOp::RedOpType::UNUSED;
+  int64_t scattered_axis = -1;
```

Comment:
It looks a bit "inconsistent" to have `scattered axis` here.

1) Why Reduce scatter needs to treated differently than other collecitves?
2) It seems to be implicitely assumed here that, for all collective, only the outer most axis can be communicated. On the one hand, this is true for now, but this will change soon. On the other hand, if we assume it for all collective, we should  consistently make the assumption for reduce scatter that the scattered axis is the second axis
3) The struct field name should reflect that it only deals with reduce scatter.
4) In general the "Communication" IR should take the TensorView pointers as input. The sharding information is all embedded there through the axis parallel types 

I understand though that some of the above comments are not due to this pr and apply to the current code.
I'm all in for simple temporary solution, so the current one might be good enough -- but I'm happy to discuss in order to align on the direction

Reply:
I had the same question, but I chose to not pulling too many changes in one PR :) 

I got the impression that there's a nccl/ProcessGroup limitation so the scattered axis has to be outermost so the scattered axis can simply be hardcoded? 

If that limitation doesn't exist or isn't fundamental, then should we support non-outermost reduce axis (which is currently hardcoded to be 0)?

Reply:
The scattered axis just has to be the outermost _allocated_ axis. Since the reduction axis isn't allocated in the output, the scattered axis can be either 0 or 1 depending on the reduction axis's placement. 

We could force it to always be the outermost axis in the RFactor domain, but that introduces unnecessary permutes